### PR TITLE
Bit dendrite debug

### PR DESCRIPTION
**Pull request by isabella618033**
separated the forward function into 3 steps 
1. forward_prep -> does all the serialization and generate a request
2. forward_call -> make the grpc call as future
3. forward_collect -> collect future and does the deserialization

The parameters was passed using a simplenamespace "forward_reques" to connect all 3 components

### CODE REVIEW

File: bittensor/_dendrite/dendrite_impl.py [hunk 1]
```diff
@@ -28,6 +28,10 @@
 from bittensor._endpoint.endpoint_impl import Endpoint
 import bittensor.utils.stats as stat_utils
 import bittensor.utils.codes as codes
+import cProfile
+import re
+import pstats
+from pstats import SortKey
```
Comment (shibshib) [inline bittensor/_dendrite/dendrite_impl.py]: Just making sure. do we need the profiler? Or is this simply for debugging purposes?

Comment (shibshib) [reply bittensor/_dendrite/dendrite_impl.py]: If we do need it, we should add it to `requirements.txt`

File: bittensor/_dendrite/dendrite_impl.py [hunk 2]
```diff
@@ -494,6 +498,10 @@ def forward_text(
                     times (:obj:`torch.FloatTensor` of shape :obj:`[ num_endpoints ]`, `required`):
                         times per call.
         """
+
+        
+        profiler = cProfile.Profile()
+        profiler.enable()
```
Comment (shibshib) [inline bittensor/_dendrite/dendrite_impl.py]: Same comments as above.

File: bittensor/_receptor/__init__.py [hunk 3]
```diff
@@ -37,12 +36,6 @@ def __new__( cls, endpoint: 'bittensor.Endpoint', wallet: 'bittensor.Wallet' = N
 
         if wallet == None:
             wallet = bittensor.wallet()
-        try:
-            external_ip = str(net.get_external_ip())
-        except Exception:
-            pass
-        finally:
-            external_ip = None
```
Comment (shibshib) [inline bittensor/_receptor/__init__.py:45]: Thank you for removing this atrocity 😄 💀

File: bittensor/_receptor/receptor_impl.py [hunk 4]
```diff
@@ -126,36 +126,11 @@ def forward (
...
-            Args:
-                inputs (:obj:`List[torch.Tensor]` of shape :obj:`(shape)`, `required`):
-                    Single torch tensor to be sent to the remote endpoint.
-
-                modality (:obj:`bittensor.proto.Modality` of shape :obj:`(1)`, `required`):
-                    Bittensor forward modality type. Enum in [TEXT, IMAGE, TENSOR]
-
-            Returns:
-                output (:obj:`Tuple[torch.FloatTensor, torch.LongTensor]`, `required`):
-                    Result tuple from the forward call.
```
Comment (shibshib) [inline bittensor/_receptor/receptor_impl.py]: Can we bring back the docustring comments?

File: bittensor/_receptor/receptor_impl.py [hunk 5]
```diff
@@ -201,24 +176,131 @@ def backward(
...
+        call_time = clock.time() - self.forward_request.start_time
+        bittensor.logging.rpc_log(
+            axon=False, 
+            forward=True, 
+            is_response=is_response, 
+            code=self.forward_request.code, 
+            call_time=call_time, 
+            pubkey=self.endpoint.hotkey, 
+            uid = self.endpoint.uid, 
+            inputs=inputs,
```
Comment (shibshib) [inline bittensor/_receptor/receptor_impl.py]: Same here, docustring comments are gone.

**Review by shibshib — COMMENTED** [review]
Haven't finished reviewing, just left some nitpick comments.

**Review by shibshib — APPROVED** [review]

**Review by unconst — APPROVED** [review]

**Pull request merged**